goship.go

```diff
@@ -760,23 +791,216 @@ func extractOutputHandler(fn func(http.ResponseWriter, *http.Request, string, st
+	if err != nil {
+		panic(err)
+	}
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		state := gomniauth.NewState("after", "success")
+
+		authUrl, err := provider.GetBeginAuthURL(state, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		http.Redirect(w, r, authUrl, http.StatusFound)
+	}
+}
+
+func callbackHandler(providerName string, auth bool) http.HandlerFunc {
+
+	if auth != true {
+		return func(w http.ResponseWriter, r *http.Request) {}
+	}
+
+	provider, err := gomniauth.Provider(providerName)
+	if err != nil {
+		panic(err)
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		omap, err := objx.FromURLQuery(r.URL.RawQuery)
+		if err != nil {
+			log.Printf("error getting resp from callback")
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		creds, err := provider.CompleteAuth(omap)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		user, userErr := provider.GetUser(creds)
+		if userErr != nil {
+			log.Printf("Failed to get user from Github %s", user)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		session, err := store.Get(r, sessionName)
+		if err != nil {
+			log.Fatal("Can't get a session store")
+		}
+
+		session.Options = &sessions.Options{
+			Path:     "/",
+			MaxAge:   86400 * 7,
+			HttpOnly: true,
+		}
+
+		session.Values["userName"] = user.Nickname()
+		session.Values["avatarURL"] = user.AvatarURL()
+		session.Save(r, w)
+
+		http.Redirect(w, r, os.Getenv("GITHUB_CALLBACK_URL"), http.StatusFound)
+	}
+}
+
+type auth struct {
+	authorization       bool
+	githubRandomHashKey string
+	githubOmniauthID    string
+	githubOmniauthKey   string
+	githubCallbackURL   string
+}
+
+//  Authenticate with Github. If env data is missing turn Auth off.
+func getAuth() auth {
+	a := auth{}
+	a.authorization = true
+	a.githubRandomHashKey = os.Getenv("GITHUB_RANDOM_HASH_KEY")
+	a.githubOmniauthID = os.Getenv("GITHUB_OMNI_AUTH_ID")
+	a.githubOmniauthKey = os.Getenv("GITHUB_OMNI_AUTH_KEY")
+	a.githubCallbackURL = os.Getenv("GITHUB_CALLBACK_URL")
+	// Let user know if a key is missing and that auth is disabled.
+	if a.githubRandomHashKey == "" || a.githubOmniauthID == "" || a.githubOmniauthKey == "" || a.githubCallbackURL == "" {
+		log.Printf("Missing one or more Gomniauth Environment Variables: Running with with limited functionality! \n githubRandomHashKey [%s] \n githubOmniauthID [%s] \n githubOmniauthKey[%s] \n githubCallbackURL[%s]",
+			a.githubRandomHashKey,
+			a.githubOmniauthID,
+			a.githubOmniauthKey,
+			a.githubCallbackURL)
+		a.authorization = false
+	} else {
+		gomniauth.SetSecurityKey(a.githubRandomHashKey)
+		gomniauth.WithProviders(
+			githubOauth.New(a.githubOmniauthID, a.githubOmniauthKey, a.githubCallbackURL+"/auth/github/callback"),
+		)
+		a.authorization = true
+	}
+	return a
```

Comment:
@ramen74 we usually try to avoid if/else blocks which end in a return and prefer to write them with two return statements like:
```go
	if a.githubRandomHashKey == "" || a.githubOmniauthID == "" || a.githubOmniauthKey == "" || a.githubCallbackURL == "" {
		log.Printf("Missing one or more Gomniauth Environment Variables: Running with with limited functionality! \n githubRandomHashKey [%s] \n githubOmniauthID [%s] \n githubOmniauthKey[%s] \n githubCallbackURL[%s]",
			a.githubRandomHashKey,
			a.githubOmniauthID,
			a.githubOmniauthKey,
			a.githubCallbackURL)
		a.authorization = false
		return a
	}
	gomniauth.SetSecurityKey(a.githubRandomHashKey)
	gomniauth.WithProviders(
		githubOauth.New(a.githubOmniauthID, a.githubOmniauthKey, a.githubCallbackURL+"/auth/github/callback"),
	)
	a.authorization = true
	return a
```